garden_ai/globus_compute/containers.py

```diff
@@ -0,0 +1,86 @@
+from time import sleep
+from enum import Enum
+from rich.console import Console
+
+from funcx import FuncXClient, ContainerSpec  # type: ignore
+from globus_sdk import GlobusAPIError
+from garden_ai.pipelines import Pipeline
+
+console = Console()
+
+
+class ContainerBuildException(Exception):
+    """Exception raised when a container build request fails"""
+
+    pass
+
+
+class BuildStatus(str, Enum):
+    queued = "queued"
+    building = "building"
+    ready = "ready"
+    failed = "failed"
+
+
+def build_container(funcx_client: FuncXClient, pipeline: Pipeline) -> str:
+    name = str(pipeline.uuid)
+    cs = ContainerSpec(
+        name=name,
+        pip=[],  # pipeline.pip_requirements,
+        python_version="3.7",  # pipeline.python_version,
+        conda=[],  # pipeline.conda_requirements,
+    )
+
+    try:
+        container_uuid = funcx_client.build_container(cs)
+    except Exception as e:
+        raise ContainerBuildException(
+            "Could not submit build request to Container Service"
+        ) from e
+
+    poll_until_container_is_built(funcx_client, container_uuid)
+    return container_uuid
+
+
+def poll_until_container_is_built(funcx_client, container_uuid):
+    """
+    Given a uuid of a container, blocks until that container is built.
+    Prints a cool status indicator to the console while polling.
+    """
+    timeout_at = 1800
+    i = 0
+    with console.status(
+        "[bold green]Building container. This operation times out after 30 minutes."
+    ) as status:
+        while i < timeout_at:
+            try:
+                status = funcx_client.get_container_build_status(container_uuid)
+            except GlobusAPIError as e:
+                raise ContainerBuildException(
+                    "Lost connection with Container Service during build"
+                ) from e
+            # Update the end user twice a minute
+            if i % 30 == 0:
+                console.log(f"Current status is {status}")
+            if status in [BuildStatus.ready, BuildStatus.failed]:
+                break
+            sleep(5)
+            i += 5
+        else:
```

Comment:
the while/else construct might be a little unclear here bc it's not super common, especially since there are a few different levels of indentation in the while loop itself -- maybe add a line comment clarifying that this is only reached when `i`  reaches `timeout_at` without having broken from the loop? Could also just do it as another if statement in the body of the loop, but imo this is the only kind of loop where the while/else construct really makes sense and it's exciting to see one in the wild

Reply:
Yeah this function is a little gnarly. I'll at least add that comment and better yet maybe just find a clearer way to write this.

Digression: Something I was thinking about but didn't pull the trigger on is that this is super old-school "async" code. (ie, not async at all. Looping and sleeping.) As we start integrating funcx calls we're going to be handling Futures and dealing with real Python async stuff. I'll be curious if we can convert this kind of code to more modern async patterns. I don't really know what that looks like in Python yet, but if this were Go there would be a channel or two here.